tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/otextension/BristolOtReceiver.java

```diff
@@ -0,0 +1,134 @@
+  private StrictBitVector choices;
+  // Index of the current random OT to use
+  private int offset = -1;
+
+  /**
+   * Initializes the underlying Rot functionality using {@code rotReceiver}. It
+   * will then construct OTs in batches of {@code batchSize}.
+   *
+   * @param rotReceiver
+   *          The random OT receiver instance to use
+   * @param resources
+   *          The resource pool for this specific instance
+   * @param network
+   *          The network to use
+   * @param batchSize
+   *          Size of the OT extension batch the protocol will construct
+   */
+  public BristolOtReceiver(RotReceiver rotReceiver,
+      OtExtensionResourcePool resources, Network network, int batchSize) {
+    this.receiver = rotReceiver;
+    this.resources = resources;
+    this.network = network;
+    this.batchSize = batchSize;
+  }
+
+  /**
+   * Receive the serialized message from the current 1-out-of-2 OT.
+   *
+   * @param choiceBit
+   *          Choice-bit. False for message 0, true for message 1.
+   * @return The serialized message from the OT
+   */
+  public byte[] receive(Boolean choiceBit) {
+    // Check if there is still an unused random OT stored, if not, execute a
+    // random OT extension
+    if (offset < 0 || offset >= batchSize) {
+      choices = new StrictBitVector(batchSize, resources.getRandomGenerator());
+      randomMessages = receiver.extend(choices);
+      offset = 0;
+    }
+    // Notify the sender if it should switch the 0 and 1 messages around (s.t.
+    // the random choice bit in the preprocessed random OTs matches the true
+    // choice bit
+    sendSwitchBit(choiceBit);
+    // Receive the serialized adjusted messages
+    byte[] zeroAdjustment = network.receive(resources.getOtherId());
+    byte[] oneAdjustment = network.receive(resources.getOtherId());
+    byte[] res = doActualReceive(zeroAdjustment, oneAdjustment);
+    offset++;
+    return res;
+  }
+
+  /**
+   * Adjust the random, preprocessed message, to fit the specific message sent
+   * by the sender.
+   *
+   * @param zeroAdjustment
+   *          The adjustment value for the zero message
+   * @param oneAdjustment
+   *          The adjustment value for the one message
+   * @return The actual message
+   */
+  private byte[] doActualReceive(byte[] zeroAdjustment, byte[] oneAdjustment) {
+    if (zeroAdjustment.length != oneAdjustment.length) {
+      throw new MaliciousException(
+          "Sender gave adjustment messages of different length.");
+    }
+    byte[] adjustment;
+    if (choices.getBit(offset, false) == false) {
+      adjustment = zeroAdjustment;
+    } else {
+      adjustment = oneAdjustment;
+    }
+    adjustMessage(adjustment);
+    return adjustment;
+  }
+
+  /**
+   * Compute and send a bit indicating whether the sender should switch the zero
+   * and one message around.
+   *
+   * @param choiceBit
+   *          The actual choice bit of the receiver
+   */
+  private void sendSwitchBit(Boolean choiceBit) {
```

Comment:
Is there a reason `choiceBit` should not be a primitive type `boolean` instead of an object `Boolean`?

Reply:
Changed in c901c59f